psi4/driver/driver.py

```diff
@@ -1931,37 +1931,135 @@ def gdma(wfn, datafile=""):
+    >>> fchk(wfn, 'mycalc.fchk')
+
+    >>> # [2] Write FCHK file with non-standard label.
+    >>> E, wfn = gradient('mp2.5', return_wfn=True)
+    >>> fchk(wfn, 'mycalc.fchk', strict_label=False)
+
     """
-    fw = core.FCHKWriter(wfn)
-    fw.write(filename)
+    # * Known limitations and notes *
+    #
+    # OCC: (occ theory module only, not dfocc) is turned off as densities are not correctly set. 
+    # DFMP2: Contains natural orbitals in wfn.C() and wfn.epsilon() data. This is fixed to contain respective HF data.
+
+    allowed = ['DFMP2', 'SCF', 'CCENERGY', 'DCT', 'DFOCC']
+    module_ = wfn.module().upper()
+    if (module_ not in allowed):
+        core.print_out(f"FCHKWriter: Theory module {module_} is currently not supported by the FCHK writer.")
+        return None
+
+    if (wfn.basisset().has_ECP()):
+        core.print_out(f"FCHKWriter: Limited ECP support! No ECP data will be written to the FCHK file.")
+
+    # fix orbital coefficients and energies for DFMP2
+    if (module_ in ['DFMP2']):
+        wfn_ = core.Wavefunction.build(wfn.molecule(), core.get_global_option('BASIS'))
+        wfn_.deep_copy(wfn)
+        refwfn = wfn.reference_wavefunction()
+        wfn_.set_reference_wavefunction(refwfn)  # refwfn not deep_copied
+        wfn_.Ca().copy(refwfn.Ca())
+        wfn_.Cb().copy(refwfn.Cb())
+        wfn_.epsilon_a().copy(refwfn.epsilon_a())
+        wfn_.epsilon_b().copy(refwfn.epsilon_b())
+        fw = core.FCHKWriter(wfn_)
+    else:
+        fw = core.FCHKWriter(wfn)
+
+    if (module_ in ['DCT', 'DFOCC']):
+        core.print_out("""FCHKWriter: Caution! For orbital-optimized correlated methods
+            the 'Orbital Energy' field contains ambiguous data. \n""")
+
+    # At this point we don't know the method name, so we try to search for it.
+    # idea: get the method from the variable matching closely the 'current energy'
+    varlist = core.scalar_variables()
+    current = varlist['CURRENT ENERGY']
+
+    # delete problematic entries
+    delete_list = ['CURRENT ENERGY', 'CURRENT REFERENCE ENERGY']
+    for key in delete_list:
+        if key in varlist:
+            del varlist[key]
+
+    # find closest matching energy
+    for (key, val) in varlist.items():
+        print(key,val)
+        if (np.isclose(val, current, 1e-12)):
+            method = key.split()[0]
+
+    # The 'official' list of labels for compatibility.
+    # OMP2,MP2.5,OCCD, etc get reduced to MP2,CC.
+    allowed_labels = {
+        "HF": " SCF Density",
+        "SCF": " SCF Density",
+        "DFT": " SCF Density",
+        "MP2": " MP2 Density",
+        "MP3": " MP3 Density",
+        "MP4": " MP4 Density",
+        "CI": " CI Density",
+        "CC": " CC Density",
+    }
+    # assign label from method name
+    fchk_label = f" {method} Density"
+    if strict_label:
+        in_list = False
+        for key in allowed_labels:
+            if key in method:
+                fchk_label = allowed_labels[key]
```

Comment:
How about a warning in case `key in method and key is not method`:
`core.print_out(f"FCHKWriter: !WARNING! {method} renamed to {key}.\n")`

Reply:
good idea